src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/service/DiagnosticService.java

```diff
@@ -0,0 +1,204 @@
+ * diagnostic service
+ */
+@Slf4j
+@Service
+@Transactional
+@PreAuthorize("isAuthenticated()")
+public class DiagnosticService {
+    private static final String DIAGNOSTICS_BASE_URL = "http://localhost";
+    private static final int CONF_CLIENT_ADMIN_PORT = SystemProperties.getConfigurationClientAdminPort();
+    private static final String CONF_CLIENT_ADMIN_PATH = "status";
+    private static final int TIMESTAMPING_SERVICE_ADMIN_PORT = PortNumbers.ADMIN_PORT;
+    private static final String TIMESTAMPING_SERVICE_ADMIN_PATH = "timestampstatus";
+    private static final int SIGNER_ADMIN_PORT = SystemProperties.getSignerAdminPort();
+    private static final String SIGNER_ADMIN_PATH = "status";
+
+    /**
+     * Query global configuration status from admin port over HTTP.
+     * @return
+     */
+    public DiagnosticsStatus queryGlobalConfStatus() {
+        try {
+            JsonObject json = sendGetRequest(buildUri(CONF_CLIENT_ADMIN_PORT, CONF_CLIENT_ADMIN_PATH));
+            return new Gson().fromJson(json, DiagnosticsStatus.class);
+        } catch (DiagnosticRequestException e) {
+            throw new DeviationAwareRuntimeException(e, e.getErrorDeviation());
+        }
+    }
+
+    /**
+     * Query timestamping services status from admin port over HTTP.
+     * @return
+     */
+    public List<DiagnosticsStatus> queryTimestampingStatus() {
+        try {
+            JsonObject json = sendGetRequest(buildUri(TIMESTAMPING_SERVICE_ADMIN_PORT,
+                    TIMESTAMPING_SERVICE_ADMIN_PATH));
+            return json.entrySet().stream().filter(e -> e.getValue() instanceof JsonObject)
+                    .map(this::parseTimestampingStatus).collect(Collectors.toList());
+        } catch (DiagnosticRequestException e) {
+            throw new DeviationAwareRuntimeException(e, e.getErrorDeviation());
+        }
+    }
+
+    /**
+     * Query ocsp responders status from admin port over HTTP.
+     * @return
+     */
+    public List<OcspResponderDiagnosticsStatus> queryOcspResponderStatus() {
+        try {
+            JsonObject json = sendGetRequest(buildUri(SIGNER_ADMIN_PORT,
+                    SIGNER_ADMIN_PATH));
+            JsonObject certificationServiceStatusMap = json.getAsJsonObject("certificationServiceStatusMap");
+            return certificationServiceStatusMap.entrySet().stream().filter(e -> e.getValue() instanceof JsonObject)
+                    .map(this::parseOcspResponderDiagnosticsStatus).collect(Collectors.toList());
+        } catch (DiagnosticRequestException e) {
+            throw new DeviationAwareRuntimeException(e, e.getErrorDeviation());
+        }
+    }
+
+    /**
+     * Send HTTP GET request to the given address (http://localhost:{port}/{path}).
+     * @param address
+     * @return
+     * @throws DiagnosticRequestException if sending a diagnostics requests fails or an error is returned
+     */
+    private JsonObject sendGetRequest(String address) throws DiagnosticRequestException {
```

Comment:
Adding a timeout is a good idea. 👍

I tried using `RestTemplate`, but in this case it does not actually simplify the implementation that much if at all. One reason for that is that the Security Server admin components return the diagnostics response using `application/octet-stream` content-type. `HttpMessageConverter` does not support that content-type and processing fails with an exception.

```
ERROR o.n.x.r.e.ApplicationExceptionHandler - exception caught
org.niis.xroad.restapi.exceptions.DeviationAwareRuntimeException: org.springframework.web.client.RestClientException: Could not extract response: no suitable HttpMessageConverter found for response type [class ee.ria.xroad.common.DiagnosticsStatus] and content type [application/octet-stream]
```

However, this can be fixed by writing a custom `HttpMessageConverter`, but it would probably just mean moving the current parsing logic to another class. Eventually it would not make the implementation much simpler compared to the additional effort required.

In addition, the current implementation utilizes the existing `DiagnosticsStatus` for all globalconf, timestamping services and ocsp responder diagnostics even if the message structures returned by the admin components are different. This makes parsing the responses little bit more complicated, but at the same time simplifies the next steps of the processing since the responses can be handled using `DiagnosticsStatus`.

Reply:
Maybe moving the logic to it's own class wouldn't be so bad 🤔 But I think this is good now since the parsing is only needed in this service.